wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt

```diff
@@ -10,95 +10,124 @@ import com.ajaxproject.warehouse.entity.MongoProduct
-        val customer: MongoCustomer = mongoCustomerRepository.findById(customerId)
-            ?: throw InternalEntityNotFoundException("Customer with id $customerId not found")
-        return WaybillDataDto(
-            id = id.toString(),
-            date = date,
-            customer = customer.mapToLiteDto(),
-            products = productList,
-            totalPrice = productList.sumOf { it.price * it.orderedAmount }
-        )
-    }
-
-    fun MongoWaybill.getListOfProducts(): List<WaybillDataDto.WaybillProductDataDto> = products.asSequence()
-        .map {
-            val product = mongoProductRepository.findById(it.productId)
-            product?.mapToWaybillProductDataDto(it.amount)
-        }
-        .filterNotNull()
-        .toList()
+    @Transactional
+    override fun updateWaybillInfo(infoUpdateDto: WaybillInfoUpdateDto, id: String): Mono<WaybillDataDto> =
+        waybillRepository.findById(ObjectId(id))
+            .switchIfEmpty { Mono.error(NotFoundException("Waybill with id $id not found")) }
+            .flatMap {
+                customerRepository.findById(ObjectId(infoUpdateDto.customerId))
+                    .switchIfEmpty {
+                        Mono.error(NotFoundException("Customer with id ${infoUpdateDto.customerId} not found"))
+                    }.thenReturn(it)
+            }.flatMap { mongoWaybill ->
+                waybillRepository
+                    .save(mongoWaybill.setUpdatedData(infoUpdateDto))
+                    .flatMap { it.mapToDataDto() }
+            }
+
+    fun MongoWaybill.mapToLiteDto(): Mono<WaybillDataLiteDto> =
+        getListOfProducts()
+            .map {
+                WaybillDataLiteDto(
+                    id = id.toString(),
+                    customerId = customerId.toString(),
+                    date = date,
+                    totalPrice = it.sumOf { it.price * it.orderedAmount }
+                )
+            }
+
+    fun MongoWaybill.mapToDataDto(): Mono<WaybillDataDto> =
+        getListOfProducts()
+            .flatMap { productList ->
+                customerRepository.findById(customerId)
+                    .switchIfEmpty {
+                        Mono.error(InternalEntityNotFoundException("Customer with id $customerId not found"))
+                    }.map {
+                        WaybillDataDto(
+                            id = id.toString(),
+                            date = date,
+                            customer = it.mapToLiteDto(),
+                            products = productList,
+                            totalPrice = productList.sumOf { it.price * it.orderedAmount }
+                        )
+                    }
+            }
+
+    fun MongoWaybill.getListOfProducts(): Mono<List<WaybillDataDto.WaybillProductDataDto>> =
+        products.asSequence()
+            .map {
+                productRepository.findById(it.productId)
```

Comment:
suppose you have 1000 products. So, you will request the DB 1000 times.

Let's rework

Reply:
done

Reply:
please check